File: app/schemas/vacancy_register.py
"""Vacancy Register -- a department-level aggregate table (additive,
read-only; no new DB table, mirrors app/schemas/reporting.py's pattern of a
Pydantic response shape over a computed aggregate query).

See app/services/vacancy_register.py for the field derivations -- none of
these (other than the plain Department passthrough fields) are literal
columns.
"""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import StaffRoleCategoryEnum
from app.schemas.common import PaginatedResponse


class VacancyRegisterRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    department_id: uuid.UUID
    department_name: str
    department_code: str | None
    supported_categories: list[StaffRoleCategoryEnum]
    is_active: bool
    campus_id: uuid.UUID
    campus_code: str

    working_count: int
    # max(approved_count - working_count, 0) -- Phase B (zany-snuggling-pie.md):
    # approved_count is now Sanctioned-Strength-backed, not
    # working_count + vacancy_count, so this can differ from the old
    # HiringSlot-derived figure.
    vacancy_count: int
    # SUM(approved_strength) across every current-effective SanctionedStrength
    # row for the department (app.services.sanctioned_strength.
    # current_effective_rows) -- Phase B: a real, independently stored
    # ceiling, no longer working_count + vacancy_count.
    approved_count: int
    # None (not 0.0) when approved_count == 0 -- there's nothing to compute a
    # percentage from, and 0.0 would misleadingly read as "confirmed zero
    # fill rate" (mirrors app/services/reporting.py's vacancy_closure_rate_pct).
    filled_pct: float | None

    requested_count: int
    approved_request_count: int
    jd_posted_count: int
    interviews_count: int
    offers_count: int
    joined_count: int

    recruitment_status: str
    # Phase B: count of VacancyRequests currently "in flight" (SUBMITTED/
    # DEAN_APPROVED/APPROVED/PUBLISHED) for this department -- see
    # app/services/vacancy_register.py's module docstring for why this is an
    # adjacent figure, not a literal "contributed to recruitment_status" count.
    recruitment_status_request_count: int
    approval_status: str
    # Phase B: count of the VacancyRequests that produced whichever
    # approval_status branch was chosen (pending count / all-time rejected
    # count / approved_request_count / 0 for NO_REQUESTS).
    approval_status_request_count: int

    last_join: date | None
    last_resignation: date | None
    last_updated: datetime


class VacancyRegisterListResponse(PaginatedResponse[VacancyRegisterRow]):
    """Additive on top of PaginatedResponse -- `items`/`total`/`limit`/
    `offset` are unchanged; `category_counts` is a snapshot of
    `{"TEACHING": n, "NON_TEACHING": n, "HOUSEKEEPING": n, "ALL": n}` across
    every active filter except `category` itself (see
    app/services/vacancy_register.py::list_vacancy_register_rows), so a
    CategoryTabs-style UI can show a live count per tab that doesn't change
    when a different tab is selected."""

    category_counts: dict[str, int]

File: frontend/src/api/vacancyRegister.ts
import { apiFetch } from "@/api/client";
import type { ApprovalStatus, PaginatedResponse, RecruitmentStatus, VacancyRegisterRow } from "@/api/types";

export type VacancyRegisterSortBy =
  | "department_name"
  | "category"
  | "approved_count"
  | "working_count"
  | "vacancy_count"
  | "filled_pct"
  | "last_join"
  | "last_resignation"
  | "last_updated";

export type SortDirection = "asc" | "desc";

export interface ListVacancyRegisterParams {
  limit?: number;
  offset?: number;
  sort_by?: VacancyRegisterSortBy;
  sort_dir?: SortDirection;
  // Filter-bar params, wired up by VacancyRegisterPage's filter UI.
  campus_code?: string | null;
  category?: string | null;
  department_id?: string | null;
  search?: string | null;
  approval_status?: ApprovalStatus | null;
  recruitment_status?: RecruitmentStatus | null;
  // Defaults server-side to true (active-only), same convention as
  // Departments/Users/Eligibility Rules -- pass false/null explicitly to see
  // inactive/all departments.
  is_active?: boolean | null;
}

// Mirrors app/schemas/vacancy_register.py::VacancyRegisterListResponse --
// additive on top of PaginatedResponse: category_counts is a snapshot of
// {"TEACHING": n, "NON_TEACHING": n, "HOUSEKEEPING": n, "ALL": n} across
// every active filter *except* category itself, so a CategoryTabs tab's
// count doesn't change when a different tab is selected.
export interface VacancyRegisterListResponse extends PaginatedResponse<VacancyRegisterRow> {
  category_counts: Record<string, number>;
}

// Returns the *full* response (not unwrapped to .items) -- unlike
// listJobPostings()/listDepartments(), this page needs `total` (and now
// category_counts) for real server-side pagination, not a
// fetch-everything-unfiltered list.
export async function listVacancyRegister(
  params: ListVacancyRegisterParams = {},
): Promise<VacancyRegisterListResponse> {
  const query = new URLSearchParams();
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));
  if (params.sort_by) query.set("sort_by", params.sort_by);
  if (params.sort_dir) query.set("sort_dir", params.sort_dir);
  if (params.campus_code) query.set("campus_code", params.campus_code);
  if (params.category) query.set("category", params.category);
  if (params.department_id) query.set("department_id", params.department_id);
  if (params.search) query.set("search", params.search);
  if (params.approval_status) query.set("approval_status", params.approval_status);
  if (params.recruitment_status) query.set("recruitment_status", params.recruitment_status);
  if (params.is_active !== undefined && params.is_active !== null) {
    query.set("is_active", String(params.is_active));
  }

  const qs = query.toString();
  return apiFetch<VacancyRegisterListResponse>(`/departments/vacancy-register${qs ? `?${qs}` : ""}`);
}
